AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/MutationStateStructures.dfy

```diff
@@ -97,27 +96,68 @@ module {:options "/functionSyntax:4" } MutationStateStructures {
     predicate ValidState()
     {
       && Commitment.Identifier == Index.Identifier
-      && 0 < |Commitment.Identifier|
       && Commitment.UUID == Index.UUID
-      && 0 < |Commitment.UUID|
+      && ValidCommitment?(Commitment)
+      && ValidIndex?(Index)
     }
+    // TODO-HV-2-FOLLOW : See if we can drop ValidUTF8
     predicate ValidUTF8()
     {
-      && UTF8.ValidUTF8Seq(Commitment.Original)
-      && UTF8.ValidUTF8Seq(Commitment.Terminal)
-      && UTF8.ValidUTF8Seq(Commitment.Input)
-      && UTF8.ValidUTF8Seq(Index.PageIndex)
-      && 0 < |Commitment.Identifier|
-      && 0 < |Index.Identifier|
-      && 0 < |Commitment.UUID|
-      && 0 < |Index.UUID|
+      && ValidCommitment?(Commitment)
+      && ValidIndex?(Index)
     }
   }
 
+  predicate ValidCommitment?(
+    Commitment: KeyStoreTypes.MutationCommitment
+  ) {
+    && UTF8.ValidUTF8Seq(Commitment.Original)
+    && UTF8.ValidUTF8Seq(Commitment.Terminal)
+    && UTF8.ValidUTF8Seq(Commitment.Input)
+    && 0 < |Commitment.Identifier|
+    && 0 < |Commitment.UUID|
+  }
+
+  predicate ValidIndex?(
+    Index: KeyStoreTypes.MutationIndex
+  ) {
+    && 0 < |Index.Identifier|
+    && 0 < |Index.UUID|
+    && UTF8.ValidUTF8Seq(Index.PageIndex)
+  }
+
+  // Pre-HV-2 Mutations did not track Hierarchy Version;
+  // therefore, HV MAY NOT be present in a Mutation Commitment.
+  // If HV is not present, we KNOW it is HV1.
+  // If HV is present, then it is the third element.
```

Comment:
(not blocking) Food for thought on putting some of the behavior in this file in spec and linking to the spec from here. But I know you're short on time.